[SYCL][FPGA] Refactor io_pipe_id attribute implementation

smanna12 opened the pull request:
This patch updates SYCL attribute: SYCLIntelPipeIOAttr to align with upstream for other attributes:
     1. stores expression as ConstantExpr in Semantic Attributes.
     2. adds AST test and more diagnostic test cases.
     3. adds merge method to handle redeclaration.
     4. handles template instantiations and duplicate attribute values properly.

Signed-off-by: Soumi Manna <soumi.manna@intel.com>

smanna12 commented:
Test failure is unrelated to my patch:
https://github.com/intel/llvm/runs/4611642482?check_suite_focus=true
Failed Tests (1):
  SYCL :: Basic/get_backend.cpp

smanna12 commented:
@elizabethandrews. could you please take a look at the patch? Thanks.

elizabethandrews commented on clang/lib/CodeGen/CodeGenModule.cpp:
```diff
@@ -4117,11 +4117,10 @@ static void maybeEmitPipeStorageMetadata(const VarDecl *D,
-    Optional<llvm::APSInt> ID =
-        IOAttr->getID()->getIntegerConstantExpr(D->getASTContext());
+  if (const auto *IOAttr = D->getAttr<SYCLIntelPipeIOAttr>()) {
+    const auto *CE = cast<ConstantExpr>(IOAttr->getID());
```
Can you explain why we need this change?

smanna12 replied:
This patch stores expression as ConstantExpr instead of IntegerConstantExpr. So i have made same changes in semantic as well as codegen part.

elizabethandrews commented on clang/test/SemaSYCL/fpga_pipes_ast.cpp:
```diff
@@ -0,0 +1,45 @@
+template <int N>
+pipe_storage Storage2 __attribute__((io_pipe_id(N)));
+
+void foo(pipe_storage PS) {}
```
What is the need for foo in this test?

smanna12 replied:
i am passing attribute value through foo(Storage2<2>);

elizabethandrews replied:
Sorry for not being more clear, but I was asking why you're doing that. Can't we just instantiate this in main ?

smanna12 replied:
Sorry i misunderstood. I see what you meant. Yes, we can instantiate this in main. Thanks for catching this.

smanna12 replied:
Done

elizabethandrews reviewed (commented):
Should we add a test to show how redeclarations are handled? From what I can tell, attributes are not inherited - is this the expected behavior?

smanna12 commented:
> Should we add a test to show how redeclarations are handled? From what I can tell, attributes are not inherited - is this the expected behavior?

Thanks for pointing this. I have added merge method to handle redeclaration. While working on it, i noticed that the attribute does not handle duplicate attribute values properly. I have fixed the problem and added test cases.

>>From what I can tell, attributes are not inherited - is this the expected behavior?
I have checked the attribute document(AttrDocs.td), i do not see in the doc that the attribute can not be inherited.

elizabethandrews commented:
> > Should we add a test to show how redeclarations are handled? From what I can tell, attributes are not inherited - is this the expected behavior?
> 
> Thanks for pointing this @elizabethandrews . I have added merge method to handle redeclaration. While working on it, i noticed that the attribute did not handle duplicate attribute values properly. I have fixed the problem and added test cases.
> 
> > From what I can tell, attributes are not inherited - is this the expected behavior?
> 
> I have checked the attribute document(AttrDocs.td), i do not see in the doc that the attribute can not be inherited.

Sorry, can you explain this? The code you added in latest commit goes against this statement right? Redeclarations will inherit attributes from existing ones now. You added a test for this as well - 

```
const pipe_storage Storage10 __attribute__((io_pipe_id(1))); // expected-note {{previous attribute is here}}
extern const pipe_storage Storage10 __attribute__((io_pipe_id(3))); // expected-warning{{attribute 'io_pipe_id' is already applied with different arguments}}
```

I do not know what this attribute does and if it even makes sense for it to be different for different declarations but if attributes cannot be inherited, I think this test is wrong. Please let me know if I am misunderstanding something

smanna12 commented:
> > > Should we add a test to show how redeclarations are handled? From what I can tell, attributes are not inherited - is this the expected behavior?
> > 
> > 
> > Thanks for pointing this @elizabethandrews . I have added merge method to handle redeclaration. While working on it, i noticed that the attribute did not handle duplicate attribute values properly. I have fixed the problem and added test cases.
> > > From what I can tell, attributes are not inherited - is this the expected behavior?
> > 
> > 
> > I have checked the attribute document(AttrDocs.td), i do not see in the doc that the attribute can not be inherited.
> 
> Sorry, can you explain this? The code you added in latest commit goes against this statement right? Redeclarations will inherit attributes from existing ones now. You added a test for this as well -
> 
> ```
> const pipe_storage Storage10 __attribute__((io_pipe_id(1))); // expected-note {{previous attribute is here}}
> extern const pipe_storage Storage10 __attribute__((io_pipe_id(3))); // expected-warning{{attribute 'io_pipe_id' is already applied with different arguments}}
> ```
> 
> I do not know what this attribute does and if it even makes sense for it to be different for different declarations but if attributes cannot be inherited, I think this test is wrong. Please let me know if I am misunderstanding something





Sorry my statement was wrong. "The attribute can not be inherited"- i do not see this anywhere in AttrDocs.td file, So my understanding is the attribute can be inherited.  


From https://github.com/intel/llvm/blob/ed4ff9a9c1b09f188c79a6202801d0bce9356b3d/clang/include/clang/Basic/AttrDocs.td#L3051-#L3063

```
def SYCLIntelPipeIOAttrDocs : Documentation {
  let Category = DocCatVariable;
  let Heading = "IO pipe ID";
  let Content = [{
Add I/O attribute to SYCL pipe declaration to declare a special I/O pipe to
interface with input or output features of an FPGA. These features might include
network interfaces, PCIe, cameras, or other data capture or processing devices
or protocols.

The io_pipe_id(id) attribute specifies the I/O feature of an accelerator board
with which a pipe interfaces. The id argument is the name of the I/O interface.
  }];
}
```

elizabethandrews commented on clang/test/SemaSYCL/fpga_pipes_ast.cpp:
```diff
@@ -0,0 +1,56 @@
+// CHECK-NEXT: ConstantExpr{{.*}}'int'
+// CHECK-NEXT: value: Int 4
+// CHECK-NEXT: IntegerLiteral {{.*}} 'int' 4
+const pipe_storage Storage3 __attribute__((io_pipe_id(4))) // expected-note{{previous attribute is here}}
```
Since you do not have `-verify ` in test line, this is not actually testing diagnostics. I don't think you need a diagnostic test here though since we already do that in another test,

If you want to keep this test, you can remove the `expected-*` and added a CHECK-NOT to ensure the conflicting attribute is not added to the AST. 

Same point for test below - maybe you can add a CHECK-NOT

smanna12 replied:
>>Since you do not have -verify in test line, this is not actually testing diagnostics. I don't think you need a diagnostic test here though since we already do that in another test,

Thanks for catching this. I have removed this test case since we already do that in another test.

>Same point for test below - maybe you can add a CHECK-NOT

I have added CHECK-NOT for this test.

elizabethandrews reviewed (commented):
I approved the patch because changes looked ok to me. However I am not sure what the required behavior is for redeclaration/inheritance. Please confirm the changes you made are functionally correct with attribute authors.

smanna12 commented:
> I approved the patch because changes looked ok to me. However I am not sure what the required behavior is for redeclaration/inheritance. Please confirm the changes you made are functionally correct with attribute authors.

Thanks @elizabethandrews for the reviews.

>>Please confirm the changes you made are functionally correct with attribute authors.

Sure.

bader commented:
@MrSidims, ping.

MrSidims commented on clang/test/SemaSYCL/fpga_pipes.cpp:
```diff
@@ -33,11 +35,51 @@ int Storage4 __attribute__((io_pipe_id(5)));
+// Test that checks template instantiations for different arg values.
+template <int A>
+pipe_storage Storage7 __attribute__((io_pipe_id(1))) // expected-note {{previous attribute is here}}
+__attribute__((io_pipe_id(A)));                      // expected-warning{{attribute 'io_pipe_id' is already applied with different arguments}}
```
I'd suggest to error out in case of redeclaration appears.

It's not expected that a common user would use this attribute. Instead it's only used on a static member (pipe storage struct) of pipe class within SYCL headers and it accepts a constexpr integer coming from pipe class template parameter as the attribute's argument. If someone modifies SYCL headers in a way of redeclaring the structure (somehow) - they should get a strong recommendation from the compiler not to do that.

Back then there was no better alternative other than an attribute. May be now we can consider removing this attribute and replacing it with implicit pipe property_list, but it goes out of a scope of this PR.

smanna12 replied:
@MrSidims, I have added error in case of redeclaration happens. Thanks for the review.

>>Back then there was no better alternative other than an attribute. May be now we can consider removing this attribute and replacing it with implicit pipe property_list, but it goes out of a scope of this PR.

Agreed. This will be of out a scope of this PR.

MrSidims reviewed (commented):
In general LGTM.

smanna12 commented:
Thanks for the reviews everyone.

@bader, This PR is ready for merge. Thank you.
